synapse/lib/remcycle.py

```diff
@@ -0,0 +1,776 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# synapse - remcycle.py
+# Created on 4/28/17.
+'''
+Remcycle provides a mechanism for kicking off asynchronous HTTP(S) requests
+via Tornado's AsyncHTTPClient.
+
+A method for templating URL endpoints, providing default configuration values
+and user set configuration values, per request, is available.  In addition,
+data can also be ingested into a cortex (provided, or a default ram cortex)
+for immediate consumption.
+
+These requests are handled by Hypnos objects, which can have a single or
+multiple definitions snapped into them in order to grab data on demand.
+Hypnos grabs these requests with a single IO thread, and offloads the
+consumption of the data to multiple worker threads.
+'''
+# Stdlib
+import cgi
+import inspect
+import logging
+import argparse
+import collections
+# Third Party Code
+import tornado.ioloop as t_ioloop
+import tornado.httpclient as t_http
+# Custom Code
+import synapse.async as s_async
+import synapse.compat as s_compat
+import synapse.cortex as s_cortex
+import synapse.lib.ingest as s_ingest
+import synapse.lib.config as s_config
+import synapse.lib.output as s_output
+import synapse.lib.threads as s_threads
+
+from synapse.common import *
+
+log = logging.getLogger(__name__)
+
+DOC = 'doc'
```

Comment:
I dont think we really have a need to create this abstraction ( the FOO='foo' ) stuff.  If the values were being dynamically calculated, and we'd save time by having static values i'd definitely do it, but e'd rather a person can see the exact value of a small static key being accessed.  And there's traditionally been a *slight* perf delta between foo.get(BAR) and foo.get('bar').